api/org/commcare/api/persistence/SqlIndexedStorageUtility.java

```diff
@@ -0,0 +1,374 @@
+
+    Class<T> prototype;
+
+    PrototypeFactory mFactory;
+
+    String tableName;
+    String userName;
+
+    public SqlIndexedStorageUtility(Class<T> prototype, PrototypeFactory factory, String userName, String tableName) {
+        this.tableName = tableName;
+        this.userName = userName;
+        this.prototype = prototype;
+        this.mFactory = factory;
+        tryCreateTable();
+    }
+
+    public Connection getConnection() throws SQLException {
+        try {
+            Class.forName("org.sqlite.JDBC");
+            return DriverManager.getConnection("jdbc:sqlite:" + this.userName + ".db");
+        } catch(Exception e){
+            System.out.println("couldn't get jdbc sqlite driver");
+        }
+        return null;
+    }
+
+    /* (non-Javadoc)
+    * @see org.javarosa.core.services.storage.IStorageUtility#write(org.javarosa.core.services.storage.Persistable)
+    */
+    public void write(Persistable p) throws StorageFullException {
+        if(p.getID() != -1) {
+            update(p.getID(), p);
+            return;
+        }
+
+        Connection c = null;
+        try {
+            c = getConnection();
+            int id = SqlHelper.insertToTable(c, tableName, p);
+            c.close();
+
+            c = getConnection();
+            p.setID(id);
+            SqlHelper.updateId(c, tableName, p);
+            c.close();
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public T readFromBytes(byte[] mBytes){
+
+        T t = null;
+        try {
+            t = prototype.newInstance();
+            ByteArrayInputStream mByteStream = new ByteArrayInputStream(mBytes);
+            t.readExternal(new DataInputStream(mByteStream), PrototypeManager.getDefault());
+            return t;
+        } catch (InstantiationException e) {
+            e.printStackTrace();
+        } catch (IllegalAccessException e) {
+            e.printStackTrace();
+        } catch (DeserializationException e) {
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /* (non-Javadoc)
+    * @see org.javarosa.core.services.storage.IStorageUtility#read(int)
+    */
+    public T read(int id) {
+        byte[] mBytes = readBytes(id);
+        return readFromBytes(mBytes);
+
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtilityIndexed#getIDsForValue(java.lang.String, java.lang.Object)
+     */
+    public Vector getIDsForValue(String fieldName, Object value) {
+
+        try {
+            Connection c = this.getConnection();
+            ResultSet rs = SqlHelper.selectFromTable(c, this.tableName,
+                    new String[]{fieldName}, new String[]{(String) value}, prototype.newInstance());
+            Vector<Integer> ids = new Vector<Integer>();
+            while(rs.next()){
```

Comment:
Please run the linter, it will warn you about important things, like the possibility of `rs.next()` throwing a `NullPointerException`